src/ring_theory/integral_domain.lean

```diff
@@ -0,0 +1,119 @@
+  (by simp [units.ext_iff.symm, subtype.coe_ext.symm])
+
+instance subgroup_units_cyclic : is_cyclic S :=
+by haveI := classical.dec_eq R; exact
+is_cyclic_of_card_pow_eq_one_le
+  (λ n hn, le_trans (card_nth_roots_subgroup_units S hn 1) (card_nth_roots _ _))
+
+end
+
+
+section
+variables {G : Type*} {R : Type*} [group G] [integral_domain R]
+
+open_locale big_operators nat
+
+open finset
+
+lemma card_fiber_eq_of_mem_range [fintype G]
+  {H : Type*} [group H] [decidable_eq H] (f : G →* H) {x y : H}
+  (hx : x ∈ set.range f) (hy : y ∈ set.range f) :
+  (univ.filter $ λ g, f g = x).card = (univ.filter $ λ g, f g = y).card :=
+begin
+  rcases hx with ⟨x, rfl⟩,
+  rcases hy with ⟨y, rfl⟩,
+  refine card_congr (λ g _, g * x⁻¹ * y) _ _ (λ g hg, ⟨g * y⁻¹ * x, _⟩),
+  { simp only [mem_filter, one_mul, monoid_hom.map_mul, mem_univ, mul_right_inv,
+      eq_self_iff_true, monoid_hom.map_mul_inv, and_self, forall_true_iff] {contextual := tt} },
+  { simp only [mul_right_inj, imp_self, forall_2_true_iff], },
+  { simp only [true_and, mem_filter, mem_univ] at hg,
+    simp only [hg, mem_filter, one_mul, monoid_hom.map_mul, mem_univ, mul_right_inv,
+      eq_self_iff_true, exists_prop_of_true, monoid_hom.map_mul_inv, and_self,
+      mul_inv_cancel_right, inv_mul_cancel_right], }
+end
+
+variables {G} [fintype G]
+
+/-- In an integral domain, a sum indexed by a nontrivial homomorphism from a finite group is zero. -/
+lemma sum_hom_units_eq_zero (f : G →* R) (hf : f ≠ 1) : ∑ g : G, f g = 0 :=
+begin
+  classical,
+  obtain ⟨x, hx⟩ : ∃ x : set.range f.to_hom_units, ∀ y : set.range f.to_hom_units, y ∈ powers x,
+    from is_cyclic.exists_monoid_generator (set.range (f.to_hom_units)),
+  have hx1 : x ≠ 1,
+  { rintro rfl,
+    apply hf,
+    ext g,
+    rw [monoid_hom.one_apply],
+    cases hx ⟨f.to_hom_units g, g, rfl⟩ with n hn,
+    rwa [subtype.coe_ext, units.ext_iff, subtype.coe_mk, monoid_hom.coe_to_hom_units,
+      is_submonoid.coe_pow, units.coe_pow, is_submonoid.coe_one, units.coe_one,
+      _root_.one_pow, eq_comm] at hn, },
+  replace hx1 : (x : R) - 1 ≠ 0,
+    from λ h, hx1 (subtype.eq (units.ext (sub_eq_zero.1 h))),
+  let c := (univ.filter (λ g, f.to_hom_units g = 1)).card,
+  calc ∑ g : G, f g
+      = ∑ g : G, f.to_hom_units g : rfl
+  ... = ∑ u : units R in univ.image f.to_hom_units, (univ.filter (λ g, f.to_hom_units g = u)).card • u :
+    sum_comp (coe : units R → R) f.to_hom_units
+  ... = ∑ u : units R in univ.image f.to_hom_units, c • u : sum_congr rfl (λ u hu, congr_arg2 _ _ rfl)
+  ... = ∑ b : set.range f.to_hom_units, c • ↑b : finset.sum_subtype
+      (by simp only [mem_image, set.mem_range, forall_const, iff_self, mem_univ, exists_prop_of_true]) _
+  ... = c • ∑ b : set.range f.to_hom_units, (b : R) : smul_sum.symm
+  ... = c • 0 : congr_arg2 _ rfl _
+  ... = 0 : smul_zero _,
+  { apply card_fiber_eq_of_mem_range f.to_hom_units,
```

Comment:
I find using `calc` like `refine` where new goals can pop up anywhere a bit hard to follow. If it's not too much work could you restructure this so that it's more clear when new goals appear and when they get proved?

Reply:
I would love to have a `defer_till_later` thingy that would behave the same like `_`.
I guess I can at least add comments, and `show`.

Reply:
Done